# Trabajo Práctico 1 - Número Pi


### Integrantes:
   Ariel Mirra
   <br>
   Juan Gabriel Ricci

### Resumen
Asdasd

### Introducción
El trabajo práctico número 1 de análisis matemático 4 se basa en una comparación de algoritmos para calcular el número matemático π. A continuación se describen los algoritmos a analizar:

 - El número π propuesto por la librería _math_ de python


### Ejercicio a
Numero π propuesto por la libreria math de Python.

In [13]:
import math

print(math.pi)

3.141592653589793


#### Descripción
Este simple ejemplo muestra el valor retornado por la librería math de Python, la librería oficial y la más utilizada.

### Ejercicio b
Algoritmo de Gauss-Legendre usando el tipo float de Python.


#### Descripción
El Algoritmo de Gauss-Legendre define variables a, b, t y p con un valor inicial, y define operaciones para realizarle a dichas variables en cada iteración. Por último, el número pi se calcula utilizando estas variables en una fórmula. Cada iteración le agrega el doble de precisión al número pi y el rendimiento del algoritmo es bastante bueno.

Al utilizar el tipo float de Python, la precisión que se puede representar es limitada como se puede mostrar en el ejemplo.

In [17]:
import math

def square(x):return x*x

def gauss_legendre(amount):
    a, b, t, p = 1, 1/math.sqrt(2), 1/4, 1
    
    for i in range(1,amount+1):
        old_a = a
        a = (a+b)/2
        b = math.sqrt(old_a*b)
        t -= p * square((old_a - a))
        p = 2 * p
        
        pi = square((a+b)) / (4*t)
        print(pi)

        
gauss_legendre(3)

3.1405792505221686
3.141592646213543
3.141592653589794


### Ejercicio c
Idem punto b, usando el tipo Decimal de Python

#### Descripción
El Algoritmo de Gauss-Legendre define variables a, b, t y p con un valor inicial, y define operaciones para realizarle a dichas variables en cada iteración. Por último, el número pi se calcula utilizando estas variables en una fórmula. Cada iteración le agrega el doble de precisión al número pi y el rendimiento del algoritmo es bastante bueno.

En este caso se implementó el algoritmo utilizando la librería Decimal de Python, que permite una mayor precisión e inlcuso definir la cantidad de decimales que se quiere. En el ejemplo a continuación se decidió utilizar 50 decimales de precisión.

In [29]:
import math
import decimal

def square(x):return x*x

def gauss_legendre(amount, precision):
    D = decimal.Decimal
    with decimal.localcontext() as ctx:
        
        ctx.prec = precision
        a, b, t, p = 1, 1/D(2).sqrt(), 1/D(4), 1
        for i in range(1,amount+1):
            old_a = a
            a = (a + b) / 2
            b = (old_a*b).sqrt()
            t -= p * square((old_a - a))
            p = 2 * p
            pi = square((a+b)) / (4*t)
            print(pi)

gauss_legendre(3, 50)

3.1405792505221682483113312689758233117734402375128
3.1415926462135422821493444319826957743144372233456
3.1415926535897932382795127748018639743812255048355


### Ejercicio d
Realizar un algoritmo de Espita (Spigot) que permita calcular los n primeros digitos de π

#### Descripción
El algoritmo de Spigot es utilizado para el calculo de digitos de π (Pi) o e (Euler). Puede calcular cantidad de digitos como se desee.
El algoritmo parte de una serie descubierta por Euler (ecuacion a continuacion). Cuando la mostramos en la forma de Horner vemos que a diferencia de π (Pi), la serie se encuentra en base 2. Gracias a este principio se puede delucidar que hara el algoritmo.

In [22]:
from IPython.display import Latex
Latex(r"""\begin{eqnarray}
      2\sum_{0}^{\infty } \frac{2^n(n!)^2}{(2n+1)!}
\end{eqnarray}""")

<IPython.core.display.Latex object>

In [19]:
def spigot_pi(n):
    k, a, b, a1, b1 = 2, 4, 1, 12, 4
    while n>0:
        p, q, k = k*k, 2*k+1, k+1
        a, b, a1, b1 = a1, b1, p*a+q*a1, p*b+q*b1
        d, d1 = a/b, a1/b1
        while d == d1:
            yield int(d)
            n -= 1
            a, a1 = 10*(a%b), 10*(a1%b1)
            d, d1 = a/b, a1/b1

print(str(eval(''.join(str(e) for e in list(spigot_pi(3)))) * (10**-2)))
print(str(eval(''.join(str(e) for e in list(spigot_pi(10)))) * (10**-10)))
print(str(eval(''.join(str(e) for e in list(spigot_pi(30)))) * (10**-29)))
print(''.join(str(e) for e in list(spigot_pi(30))))

3.14
3.1415926535
31.41592653589793
3141592653589793238462643383279


### Analisis del código
Para el análisis de los diferentes algoritmos e implementaciones, vamos a comparar los resultados obtenidos contra el valor otorgado por la librería math de Python:
#### π = 3.141592653589793
---

#### Gauss Legendre
El algoritmo de Gauss-Legendre se realizó en ambos casos con tres iteraciones, obteniendo los siguientes resultados:
##### utilizando float:
* 3.1405792505221686
* 3.141592646213543
* 3.141592653589794

##### utilizando decimal:
* 3.1405792505221682483113312689758233117734402375128
* 3.1415926462135422821493444319826957743144372233456
* 3.1415926535897932382795127748018639743812255048355

Como se pudo analizar, al utilizar float, Python redondea en el dígito número 15 después de la coma, por lo que la máxima precisión que se puede obtener son 14 dígitos. Sin embargo, dicha precisión recién se logra en la tercera iteración del algoritmo.

Utilizando Decimal, se pueden obtener más dígitos de precisión. Sin embargo, recién se logra mayor precisión que el número otorgado por la librería math en la tercera iteración. Decimal nos permite obtener mucha más precisión en caso de que se necesite.